SyntaxWarning: 'return' in a 'finally' block

mcatanzaro opened the pull request:
#### 622fea258aae0e8d4eb8489dcd78893ca06dd494
<pre>
SyntaxWarning: &apos;return&apos; in a &apos;finally&apos; block
<a href="https://bugs.webkit.org/show_bug.cgi?id=297685">https://bugs.webkit.org/show_bug.cgi?id=297685</a>

Reviewed by Sam Sneddon.

Return in a finally block is now discouraged, see: <a href="https://peps.python.org/pep-0765/">https://peps.python.org/pep-0765/</a>

Much of this is copied from a suggestion by Sam Sneddon.

Canonical link: <a href="https://commits.webkit.org/299064@main">https://commits.webkit.org/299064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81c6413a4033564b28c17416da42892a4f577e8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117667 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123785 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69678 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119545 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45927 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89296 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/49913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120619 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105503 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69786 "1 api test failed or timed out") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/117026 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29342 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23619 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67457 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99696 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23799 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126894 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44570 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33542 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97962 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/117087 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44928 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101730 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97749 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24875 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43127 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21086 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40934 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44441 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43900 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47247 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45591 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/terminal.py:
```diff
@@ -316,7 +316,8 @@ def apply(self, target):
                     yield
                 finally:
-                    return
+                    pass
```
A `finally` block with only `pass` in it does nothing, so we can remove the entire `try .. finally`.

mcatanzaro replied:
That leaves me with:

```
            try:
                previous = self.top.get(target.fileno())
            except (io.UnsupportedOperation, AttributeError):
                yield
                return
```

I am insufficiently familiar with generators and I'm wondering whether this actually makes sense? Why not just:

```
            try:
                previous = self.top.get(target.fileno())
            except (io.UnsupportedOperation, AttributeError):
                return
```

gsnedders replied:
It does make sense — the continuation yields, and then when the continuation resumes it will then return.

```python
import contextlib

@contextlib.contextmanager
def cm_try_finally():
    try:
        print("before yield")
        yield
        print("after yield")
    finally:
        print("before return")
        return
        print("after return")


@contextlib.contextmanager
def cm_yield_return():
    print("before yield")
    yield
    print("after yield")
    print("before return")
    return
    print("after return")


print("cm_try_finally: empty with statement:")
with cm_try_finally():
    pass

print()
print("cm_try_finally: raise in with statement:")
with cm_try_finally():
    raise Exception()


print()
print("cm_yield_return: empty with statement:")
with cm_yield_return():
    pass

print()
print("cm_yield_return: raise in with statement:")
with cm_yield_return():
    raise Exception()
```

Outputs:

```
cm_try_finally: empty with statement:
before yield
after yield
before return

cm_try_finally: raise in with statement:
before yield
before return

cm_yield_return: empty with statement:
before yield
after yield
before return

cm_yield_return: raise in with statement:
before yield
Traceback (most recent call last):
  File "/tmp/foo.py", line 43, in <module>
    raise Exception()
Exception
```

Which shows the prior significance of the `finally` block.

So I think what we want is:

```python
try:
    previous = self.top.get(target.fileno())
except (io.UnsupportedOperation, AttributeError):
    try:
        yield
    except Exception:
        pass
    return
```

mcatanzaro replied:
Thanks. I've updated this pull request.

At this point, you've written more of the pull request than I did, so I've added a credit in the commit message.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
```diff
@@ -168,11 +168,13 @@ def _append(branch, hash, revision=None):
+            if log and log.returncode:
+                return
+            if log and log.poll() is None:
+                log.kill()
```
The first `if` statement is wrong, but that's a pre-existing bug — `log.returncode` will always be `None` except in the case where we've raised `self.repo.Exception` above ([`Popen.returncode`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.returncode) is only set by `poll`, `wait`, and `communicate`).

For the second `if` statement, moving it out of the `finally` block means we won't actually run it if the `try` block raises, which is definitely wrong.

I think what we want here is `if log and log.poll(): return` at the end of the `try` block, and then the `if log and log.poll() is None: log.kill()` still in the finally block.

mcatanzaro committed:
SyntaxWarning: 'return' in a 'finally' block
https://bugs.webkit.org/show_bug.cgi?id=297685

Reviewed by Sam Sneddon.

Return in a finally block is now discouraged, see: https://peps.python.org/pep-0765/

Much of this is copied from a suggestion by Sam Sneddon.

Canonical link: https://commits.webkit.org/299064@main
